Summarizing or sorting in query editor after Exploring results in a SQL question and creating a Custom Column breaks query editor

### Describe the bug

After creating a SQL query, clicking explore results, creating a custom column, and then clicking Summarize or Order By, the GUI editor will break with "Something's gone wrong. We've run into an error. You can try refreshing the page, or just go back."

Hopefully, this will help us understand the root of this error that has been more frequent in the last couple of versions.

### To Reproduce

1. Create a SQL question based on Sample Database: select * from orders
2. Save it and click "Explore Results" top right
3. Go to GUI editor and create a custom column: [TAX] --a test column
4. Click Sort or Summarize, it should bring the error "Something's gone wrong"

### Expected behavior

_No response_

### Logs

I don't know if this helps, but here are console logs:
```
react-dom.production.min.js:209 RangeError: Maximum call stack size exceeded
    at e (cljs.core.js:500:178)
    at cljs.core.js:501:57
    at e (cljs.core.js:501:3)
    at cljs.core.js:501:57
    at e (cljs.core.js:501:3)
    at cljs.core.js:501:57
    at e (cljs.core.js:501:3)
    at yr.h.ya (cljs.core.js:505:37)
    at F (cljs.core.js:22:45)
    at yr.Em.h (cljs.core.js:248:369)
cs @ react-dom.production.min.js:209
n.payload @ react-dom.production.min.js:227
mo @ react-dom.production.min.js:129
Ua @ react-dom.production.min.js:179
xs @ react-dom.production.min.js:263
Sc @ react-dom.production.min.js:246
_c @ react-dom.production.min.js:246
gc @ react-dom.production.min.js:239
(anonymous) @ react-dom.production.min.js:123
t.unstable_runWithPriority @ scheduler.production.min.js:19
Vi @ react-dom.production.min.js:122
Ji @ react-dom.production.min.js:123
Zi @ react-dom.production.min.js:122
T @ react-dom.production.min.js:287
Kt @ react-dom.production.min.js:68
app.ts:22 Error: {componentStack: '\n    in xn\n    in div\n    in u\n    in span\n    in …\n    in h\n    in DragDropContextProvider\n    in I'}
h @ app.ts:22
n.<computed> @ bindActionCreators.js:8
value @ ErrorBoundary.tsx:31
o.componentDidCatch.n.callback @ react-dom.production.min.js:227
Mo @ react-dom.production.min.js:131
hs @ react-dom.production.min.js:212
Ec @ react-dom.production.min.js:255
t.unstable_runWithPriority @ scheduler.production.min.js:19
Vi @ react-dom.production.min.js:122
Rc @ react-dom.production.min.js:248
gc @ react-dom.production.min.js:239
(anonymous) @ react-dom.production.min.js:123
t.unstable_runWithPriority @ scheduler.production.min.js:19
Vi @ react-dom.production.min.js:122
Ji @ react-dom.production.min.js:123
Zi @ react-dom.production.min.js:122
T @ react-dom.production.min.js:287
Kt @ react-dom.production.min.js:68
StructuredQuery.ts:368 Warning: can't clean query without metadata!
value @ StructuredQuery.ts:368
value @ Question.ts:1143
f @ urls.ts:42
P @ questions.ts:106
(anonymous) @ NewItemMenu.tsx:79
useMemo @ react-dom.production.min.js:165
t.useMemo @ react.production.min.js:25
Bi @ NewItemMenu.tsx:72
ta @ react-dom.production.min.js:153
xs @ react-dom.production.min.js:261
Sc @ react-dom.production.min.js:246
_c @ react-dom.production.min.js:246
gc @ react-dom.production.min.js:239
(anonymous) @ react-dom.production.min.js:123
t.unstable_runWithPriority @ scheduler.production.min.js:19
Vi @ react-dom.production.min.js:122
Ji @ react-dom.production.min.js:123
Zi @ react-dom.production.min.js:122
T @ react-dom.production.min.js:287
Kt @ react-dom.production.min.js:68
[Violation] 'click' handler took 223ms
```
### Information about your Metabase installation

JSON
47.2



### Severity

very specific steps, but can be annoying 

### Additional context

_No response_

--- a/src/metabase/lib/field.cljc
+++ b/src/metabase/lib/field.cljc
@@ -77,24 +77,38 @@
                             (pr-str (mapv :lib/desired-column-alias column-metadatas))))
         nil)))
 
+(def ^:private ^:dynamic *recursive-column-resolution-by-name*
+  "Whether we're in a recursive call to [[resolve-column-name]] or not. Prevent infinite recursion (#32063)"
+  false)
+
 (mu/defn ^:private resolve-column-name :- [:maybe lib.metadata/ColumnMetadata]
   "String column name: get metadata from the previous stage, if it exists, otherwise if this is the first stage and we
   have a native query or a Saved Question source query or whatever get it from our results metadata."
   [query        :- ::lib.schema/query
    stage-number :- :int
    column-name  :- ::lib.schema.common/non-blank-string]
-  (let [stage         (if-let [previous-stage-number (lib.util/previous-stage-number query stage-number)]
-                        (lib.util/query-stage query previous-stage-number)
-                        (lib.util/query-stage query stage-number))
-        ;; TODO -- it seems a little icky that the existence of `:metabase.lib.stage/cached-metadata` is leaking here,
-        ;; we should look in to fixing this if we can.
-        stage-columns (or (:metabase.lib.stage/cached-metadata stage)
-                          (get-in stage [:lib/stage-metadata :columns])
-                          (when (:source-card stage)
-                            (lib.metadata.calculation/visible-columns query stage-number stage))
-                          (log/warn (i18n/tru "Cannot resolve column {0}: stage has no metadata" (pr-str column-name))))]
-    (when (seq stage-columns)
-      (resolve-column-name-in-metadata column-name stage-columns))))
+  (when-not *recursive-column-resolution-by-name*
+    (binding [*recursive-column-resolution-by-name* true]
+      (let [previous-stage-number (lib.util/previous-stage-number query stage-number)
+            stage                 (if previous-stage-number
+                                    (lib.util/query-stage query previous-stage-number)
+                                    (lib.util/query-stage query stage-number))
+            ;; TODO -- it seems a little icky that the existence of `:metabase.lib.stage/cached-metadata` is leaking
+            ;; here, we should look in to fixing this if we can.
+            stage-columns         (or (:metabase.lib.stage/cached-metadata stage)
+                                      (get-in stage [:lib/stage-metadata :columns])
+                                      (when (:source-card stage)
+                                        (lib.metadata.calculation/visible-columns query stage-number stage))
+                                      (log/warn (i18n/tru "Cannot resolve column {0}: stage has no metadata"
+                                                          (pr-str column-name))))]
+        (when-let [column (and (seq stage-columns)
+                               (resolve-column-name-in-metadata column-name stage-columns))]
+          (cond-> column
+            previous-stage-number (-> (dissoc :id :table-id
+                                              ::binning ::temporal-unit)
+                                      (lib.join/with-join-alias nil)
+                                      (assoc :name (or (:lib/desired-column-alias column) (:name column)))
+                                      (assoc :lib/source :source/previous-stage))))))))
 
 (mu/defn ^:private resolve-field-metadata :- lib.metadata/ColumnMetadata
   "Resolve metadata for a `:field` ref. This is part of the implementation

--- a/test/metabase/lib/field_test.cljc
+++ b/test/metabase/lib/field_test.cljc
@@ -4,6 +4,7 @@
    [medley.core :as m]
    [metabase.lib.binning :as lib.binning]
    [metabase.lib.core :as lib]
+   [metabase.lib.field :as lib.field]
    [metabase.lib.metadata :as lib.metadata]
    [metabase.lib.metadata.calculation :as lib.metadata.calculation]
    [metabase.lib.metadata.composed-provider
@@ -684,3 +685,25 @@
              (lib.metadata.calculation/metadata
               query
               [:field {:lib/uuid "aa0e13af-29b3-4c27-a880-a10c33e55a3e", :base-type :type/Text} 4]))))))
+
+(deftest ^:parallel recursive-column-resolution-test
+  (testing "#34247"
+    (let [source-query      (lib/query meta/metadata-provider (meta/table-metadata :orders))
+          metadata-provider (lib.metadata.composed-provider/composed-metadata-provider
+                             meta/metadata-provider
+                             (lib.tu/mock-metadata-provider
+                              {:cards [{:name          "My Card"
+                                        :id            1
+                                        :dataset-query source-query}]}))
+          query             (-> (lib/query metadata-provider {:database (meta/id)
+                                                              :lib/type :mbql/query
+                                                              :stages   [{:lib/type    :mbql.stage/mbql
+                                                                          :source-card 1}]})
+                                (lib/expression "expr" [:field
+                                                        {:lib/uuid  (str (random-uuid))
+                                                         :base-type :type/Float}
+                                                        "TAX"]))]
+      (is (=? {:lib/type :metadata/column
+               :name     "TAX"
+               :id       (meta/id :orders :tax)}
+             (#'lib.field/resolve-column-name query -1 "TAX"))))))

